### PR TITLE
change(utils): Add a direct connection mode to zebra-checkpoints

### PR DESCRIPTION
**Pull request by teor2345**
## Instructions

Test the new direct JSON-RPC connections using:
```sh
cargo run --release --features=zebra-checkpoints --bin zebra-checkpoints -- --transport direct --addr 127.0.0.1:<zebrad port>
```

And the refactored existing functionality using:
```sh
cargo run --release --features=zebra-checkpoints --bin zebra-checkpoints -- --transport cli --addr 127.0.0.1:<zebrad port>
# automatically uses zcashd
cargo run --release --features=zebra-checkpoints --bin zebra-checkpoints -- --transport cli --backend zcashd
# actually uses zcashd, but in zebrad RPC parameter mode
cargo run --release --features=zebra-checkpoints --bin zebra-checkpoints -- --transport cli --backend zebrad
```

## Motivation

In #6367 we want to generate checkpoints in CI. This is the Rust tool part of that ticket.

### Complex Code or Requirements

There are 4 modes in this tool now:
- zebrad direct
- zcashd CLI
- zebrad CLI
- zcashd CLI in zebrad mode (not supported but it should work)

This means we need to test all of them. (Which I've done.)

Technically this code is async, but it doesn't actually use much concurrency.

## Solution

- [Move all the CLI-specific code into a single function](https://github.com/ZcashFoundation/zebra/commit/86b7cb291078d57ffda48259ee737eeecf7bd6d7)
- [Add transport and addr arguments to zebra-checkpoints](https://github.com/ZcashFoundation/zebra/commit/e14aab69b21b89dc158c115b8229544d86ce7c6e)
- [Make zebra-checkpoints support both CLI and direct JSON-RPC connections](https://github.com/ZcashFoundation/zebra/commit/f10a7889dc4d395d6207ed5918892519fb1a533e)
- [Document required features](https://github.com/ZcashFoundation/zebra/commit/9daf7d87654455c69c3f84a77a00b668f06e4318)

Related Changes:
- simplify the `zebra-checkpoints` implementation
- [Add a json-conversion feature for converting JSON to valid Heights](https://github.com/ZcashFoundation/zebra/commit/b9ec68bfd845d26f7fc5ff8d1465a97875d934bb)
    - Add a `TryIntoHeight` convenience trait for converting integers, strings, and JSON to valid heights (we'll use this in other fixes like #6279)
- [Move the RpcRequestClient into zebra-node-services](https://github.com/ZcashFoundation/zebra/commit/27ba4bf356d8ec8f05accfe6fd400f0138f61ee1), behind a new feature

Unrelated cleanups:
- [Replace a custom function with an existing dependency function](https://github.com/ZcashFoundation/zebra/pull/6516/commits/e89308bc561b9caaa59bca377785e6a67d99b5bd)

### Testing

- [Add a checkpoint list test for minimum height gaps](https://github.com/ZcashFoundation/zebra/pull/6516/commits/d78be65cc255504fd0b2d8e448ddd8992e594cfe)

The rest of the testing will happen as part of the next PR which puts this in CI.

## Review

This is part of a routine release process simplification change.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

Add this new tool to CI, so we always have checkpoints ready for releases.
Fully test the generated checkpoints.

### CODE REVIEW

File: zebra-utils/src/bin/zebra-checkpoints/main.rs [hunk 1]
```diff
@@ -58,87 +125,111 @@ fn cmd_output(cmd: &mut std::process::Command) -> Result<String> {
...
-        .expect("zcashd has some mature blocks: wait for zcashd to sync more blocks");
-
-    let starting_height = args::Args::from_args().last_checkpoint.map(block::Height);
-    if starting_height.is_some() {
-        // Since we're about to add 1, height needs to be strictly less than the maximum
-        assert!(starting_height.unwrap() < block::Height::MAX);
-    }
+        - HeightDiff::try_from(MIN_TRANSPARENT_COINBASE_MATURITY).expect("constant fits in i32");
+    let height_limit =
+        height_limit.expect("node has some mature blocks: wait for it to sync more blocks");
```
Comment (upbqdn) [inline zebra-utils/src/bin/zebra-checkpoints/main.rs:165]: Unfortunately, I don't understand what this error message means. The error occurs when the node has less than 100 blocks, which means it's a fresh node. What about something like the following?
```suggestion
        let height_limit = height_limit.expect(
        "The node should have more than {MIN_TRANSPARENT_COINBASE_MATURITY} synced blocks.",
    );
```

File: zebra-utils/src/bin/zebra-checkpoints/main.rs [hunk 2]
```diff
@@ -58,87 +125,111 @@ fn cmd_output(cmd: &mut std::process::Command) -> Result<String> {
...
+            .expect("invalid last checkpoint height, must be less than the max height")
+    } else {
+        Height::MIN
+    };
 
     assert!(
-        starting_height < height_limit.0,
-        "No mature blocks after the last checkpoint: wait for zcashd to sync more blocks"
+        starting_height < height_limit,
+        "No mature blocks after the last checkpoint: wait for node to sync more blocks"
```
Comment (upbqdn) [inline zebra-utils/src/bin/zebra-checkpoints/main.rs:178]: ###### Optional

What about something like

```suggestion
        "The height of the last checkpoint must be less than the height of the last mature block."
```

File: zebra-utils/src/bin/zebra-checkpoints/main.rs [hunk 3]
```diff
@@ -58,87 +125,111 @@ fn cmd_output(cmd: &mut std::process::Command) -> Result<String> {
...
+                    .expect("block bytes: unexpected missing field or field type");
 
-                let block_bytes = <Vec<u8>>::from_hex(output.trim_end_matches('\n'))?;
+                let block_bytes: Vec<u8> = hex::decode(block_bytes)?;
 
-                let block = block_bytes
-                    .zcash_deserialize_into::<block::Block>()
-                    .expect("obtained block should deserialize");
+                // TODO: is it faster to call both `getblock height 0` and `getblock height 1`,
+                //       rather than deserializing the block and calculating its hash?
```
Comment (upbqdn) [inline zebra-utils/src/bin/zebra-checkpoints/main.rs:231]: Why do we mention those heights in the TODO?

File: zebra-consensus/src/checkpoint/list/tests.rs [hunk 4]
```diff
@@ -274,14 +277,21 @@ fn checkpoint_list_hard_coded_max_gap_testnet() -> Result<(), BoxError> {
...
 
-/// Check that the hard-coded checkpoints are within `MAX_CHECKPOINT_HEIGHT_GAP`.
+/// Check that the hard-coded checkpoints are within [`MAX_CHECKPOINT_HEIGHT_GAP`],
+/// and a calculated minimum number of blocks. This also checks the heights are in order.
 ///
-/// We can't test the block byte limit, because we don't have access to the entire
-/// blockchain in the tests. But that's ok, because the byte limit only impacts
-/// performance.
+/// We can't test [`MAX_CHECKPOINT_BYTE_COUNT`] directly, because we don't have access to the
+/// entire blockchain in the tests. Instead, we check the number of maximum-size blocks in a
```
Comment (upbqdn) [inline zebra-consensus/src/checkpoint/list/tests.rs:284]: ###### Optional

```suggestion
/// We can't test [`MAX_CHECKPOINT_BYTE_COUNT`] directly, because we don't have access to a large
/// enough blockchain in the tests. Instead, we check the number of maximum-size blocks in a
```

**Review by upbqdn — APPROVED** [review]
Looks great. I left a couple of minor suggestions and questions.

File: zebra-utils/src/bin/zebra-checkpoints/main.rs (hunk 1 above)
Comment (teor2345) [reply zebra-utils/src/bin/zebra-checkpoints/main.rs:165]: Thanks for this review, none of these things are blocking, so I'll fix them in a follow-up PR, to unblock the next PR in this series merging.

Comment (teor2345) [reply zebra-utils/src/bin/zebra-checkpoints/main.rs:165]: These are now fixed in https://github.com/ZcashFoundation/zebra/pull/6579/files
